### PR TITLE
Adds BRIEF Descriptor

### PR DESCRIPTION
**Pull request by mronian**

### CODE REVIEW

File: src/brief.jl [hunk 1]
```diff
@@ -0,0 +1,109 @@
+type BRIEF <: DescriptorParams
+	size::Int
+	window::Int
+	sigma::Float64
+	sampling_type::Function
```
Comment (timholy) [inline src/brief.jl]: `sampling_type::F` and make `F` a parameter

File: src/brief.jl [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	sample_two = CartesianIndex{2}[]
+	while true
+		x_gen = floor(Int, window * rand() / 2)
+		y_gen = floor(Int, window * rand() / 2)
+		(x_gen > -window / 2 && x_gen < (window - 1) / 2) && (y_gen > -window / 2 && y_gen < (window - 1) / 2) || continue
+		x_gen_2 = floor(Int, window * rand() / 2)
+		y_gen_2 = floor(Int, window * rand() / 2)
+		(x_gen_2 > -window / 2 && x_gen_2 < (window - 1) / 2) && (y_gen_2 > -window / 2 && y_gen_2 < (window - 1) / 2) || continue
+		count += 1
+		push!(sample_one, CartesianIndex{2}(y_gen, x_gen))
```
Comment (timholy) [inline src/brief.jl]: Do you need the {2}?

File: src/brief.jl [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	size::Int
+	window::Int
+	sigma::Float64
+	sampling_type::Function
+	seed::Int
+end
+
+function random_uniform(size::Int, window::Int, seed::Int)
+	srand(seed)
+	sample_one = CartesianIndex{2}[]
```
Comment (timholy) [inline src/brief.jl]: (ref below, here you definitely want the `{2}`)

File: src/brief.jl [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+		y_gen = floor(Int, window * rand() / 2)
+		(x_gen > -window / 2 && x_gen < (window - 1) / 2) && (y_gen > -window / 2 && y_gen < (window - 1) / 2) || continue
+		count += 1
+		push!(sample, CartesianIndex{2}(y_gen, x_gen))
+		count != size || break
+	end
+	zeros(CartesianIndex{2}, size), sample
+end
+
+function BRIEF(; size::Integer = 128, window::Integer = 9, sigma::Float64 = 2 ^ 0.5, sampling_type::Function = gaussian, seed::Int = 123)
```
Comment (timholy) [inline src/brief.jl]: I'd move this up to just below the type definition.

**Comment by mronian** [comment]
@timholy This is ready for review. 

The Mac build on Travis is failing due to `ImageMagick` not being built correctly. I haven't changed the code on travis so I don't know why it is failing. Could you take a look at it?

**Comment by timholy** [comment]
I don't have a mac, or even access to one---my entire family and I believe my entire lab is currently mac-free (must be some kind of record).

The last test on Travis of the package was successful, so it must be a recent change.

**Comment by mronian** [comment]
The problem with OSX was that it needs `QuratzImageIO` instead of `ImageMagick` (found this in `test/REQUIRE` of `Images.jl` :stuck_out_tongue: ). I have added tests with `lena` and `lighthouse`.

@timholy Is this good to merge? :)

File: test/runtests.jl [hunk 5]
```diff
@@ -1,10 +1,52 @@
...
+    check_bool = check_bool && length(sample_two) == size
+    for s1 in sample_one, s2 in sample_two
+        check_bool = check_bool && (s1[1] >= ceil(-window / 2) && s1[1] <= floor((window - 1) / 2)) && (s1[2] >= ceil(-window / 2) && s1[2] <= floor((window - 1) / 2))
+        check_bool = check_bool && (s2[1] >= ceil(-window / 2) && s2[1] <= floor((window - 1) / 2)) && (s2[2] >= ceil(-window / 2) && s2[2] <= floor((window - 1) / 2))
+    end
+    return check_bool 
+end
+
+function _warp(img, transx, transy)
+    res = zeros(eltype(img), size(img))
```
Comment (mronian) [inline test/runtests.jl:18]: Will move this to `CameraGeometry` once GSoC is done. It is too simple to be moved right now.

Comment (mronian) [inline test/runtests.jl:18]: Also, https://github.com/timholy/AffineTransforms.jl is better as a base for `CameraGeometry.jl`.

File: src/core.jl [hunk 6]
```diff
@@ -1,9 +1,36 @@
 abstract Detector
-  
+abstract DescriptorParams
 typealias Keypoint CartesianIndex{2}
 typealias Keypoints Array{CartesianIndex{2}}
```
Comment (timholy) [inline src/core.jl]: `Vector{CartesianIndex{2}}` or use

``` jl
typealias Keypoints{N} Array{CartesianIndex{2},N}
```

so it's a leaftype.

File: src/brief.jl [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        push!(sample, CartesianIndex(y_gen, x_gen))
+        count != size || break
+    end
+    zeros(CartesianIndex{2}, size), sample
+end
+
+function create_descriptor{T<:Gray}(img::AbstractArray{T, 2}, keypoints::Keypoints, params::BRIEF)
+    img_smoothed = imfilter_gaussian(img, [params.sigma, params.sigma])
+    sample_one, sample_two = params.sampling_type(params.size, params.window, params.seed)
+    descriptors = BitArray[]
```
Comment (timholy) [inline src/brief.jl]: `BitArray{1}`? (`BitArray` is not a leaftype.)

File: test/runtests.jl (hunk 5 above)
Comment (timholy) [inline test/runtests.jl:18]: Even that might be replaced by https://github.com/FugroRoames/CoordinateTransformations.jl (I haven't had time to look, but it seems promising).

**Comment by timholy** [comment]
LGTM!

**Pull request merged**